File: core/clientLaya/game/src/commonGame/event/SBaseEventRegister.ts
namespace Shine
{
	/** 基础事件注册 */
	export class SBaseEventRegister<T>
	{
		private _indexToTypeDic:SMap<number,number>=new SMap<number,number>();
	
		private _listenerDic:SMap<number,SMap<number,BaseEventAction<T>>> = new SMap<number,SMap<number,BaseEventAction<T>>> ();
		/** 序号 */
		private _index:number=0;
	
		public addListenerFun(type:number,listener:Func):number
		{
			var func:BaseEventAction<T>=new BaseEventAction();
			func.func=listener;
	
			return this.addListener(type,func);
		}
	
		/**添加监听 */
		public addListener(type:number,func:BaseEventAction<T>):number
		{
			var dic:SMap<number,BaseEventAction<T>> = this._listenerDic.get(type);
	
			if(dic==null)
			{
				this._listenerDic.put(type,dic = new SMap<number,BaseEventAction<T>>());
			}
	
			var index:number=++this._index;
	
			dic.put(index,func);
			this._indexToTypeDic.put(index,type);
	
			return index;
		}
	
		/** 移除监听 */
		public removeListener(type:number,index:number):void
		{
			var dic:SMap<number,BaseEventAction<T>> = this._listenerDic.get(type);
	
			if(dic==null)
			{
				return;
			}
	
			dic.remove(index);
			this._indexToTypeDic.remove(index);
		}
	
		/** 移除监听通过序号 */
		public removeListener1(index:number):void
		{
			var type:number=this._indexToTypeDic.get(index);
	
			if(type>0)
			{
				this.removeListener(type,index);
			}
		}
	
		/** 删除所有监听 */
		public removeAllListener():void
		{
			this._listenerDic.clear();
		}
	
		/** 派发消息 */
		public dispatch(type:number,data:T=null):void
		{
			var dic:SMap<number,BaseEventAction<T>> = this._listenerDic.get(type);
	
			if(dic==null)
			{
				return;
			}
			
			// var keys:number[] =dic.getKeys();
			// var values:BaseEventAction<T>[] =dic.getValues();
			// var fv:number=dic.getFreeValue();
			// var k:number;
	
			//**********不确定的**********
			dic.forEach((k,v)=>{
				v.execute(data);
			});

			// for(var i:number=keys.length-1;i>=0;--i)
			// {
			// 	if((k=keys[i])!=fv)
			// 	{
			// 		values[i].execute(data);
	
			// 		if(k!=keys[i])
			// 		{
			// 			++i;
			// 		}
			// 	}
			// }
		}

	}
}

File: core/client/game/src/commonGame/event/SBaseEventRegister.cs
using System;
using ShineEngine;

/// <summary>
/// 基础事件注册
/// </summary>
public class SBaseEventRegister<T>
{
	private IntIntMap _indexToTypeDic=new IntIntMap();

	private IntObjectMap<IntObjectMap<BaseEventAction>> _listenerDic=new IntObjectMap<IntObjectMap<BaseEventAction>>();
	/** 序号 */
	private int _index=0;

	/// <summary>
	/// 添加监听
	/// </summary>
	public int addListener(int type,Action listener)
	{
		BaseEventAction func=new BaseEventAction();
		func.func=listener;

		return addListener(type,func);
	}

	/// <summary>
	/// 添加监听
	/// </summary>
	public int addListener(int type,Action<T> listener)
	{
		BaseEventAction func=new BaseEventAction();
		func.func2=listener;

		return addListener(type,func);
	}

	/// <summary>
	/// 添加监听
	/// </summary>
	public int addListener(int type,BaseEventAction func)
	{
		IntObjectMap<BaseEventAction> dic=_listenerDic.get(type);

		if(dic==null)
		{
			_listenerDic.put(type,dic=new IntObjectMap<BaseEventAction>());
		}

		int index=++_index;

		dic.put(index,func);
		_indexToTypeDic.put(index,type);

		return index;
	}

	/// <summary>
	/// 移除监听
	/// </summary>
	public void removeListener(int type,int index)
	{
		IntObjectMap<BaseEventAction> dic=_listenerDic.get(type);

		if(dic==null)
		{
			return;
		}

		dic.remove(index);
		_indexToTypeDic.remove(index);
	}

	/// <summary>
	/// 移除监听通过序号
	/// </summary>
	public void removeListener(int index)
	{
		int type=_indexToTypeDic.get(index);

		if(type>0)
		{
			removeListener(type,index);
		}
	}

	/// <summary>
	/// 删除所有监听
	/// </summary>
	public void removeAllListener()
	{
		foreach(var v in _listenerDic)
		{
			v.clear();
		}
	}

	/// <summary>
	/// 派发消息
	/// </summary>
	public void dispatch(int type,T data=default(T))
	{
		IntObjectMap<BaseEventAction> dic=_listenerDic.get(type);

		if(dic==null)
		{
			return;
		}

		int[] keys=dic.getKeys();
		BaseEventAction[] values=dic.getValues();
		int fv=dic.getFreeValue();
		int k;
		int safeIndex=dic.getLastFreeIndex();

		for(int i=safeIndex - 1;i!=safeIndex;--i)
		{
			if(i<0)
			{
				i=values.Length;
			}
			else if((k=keys[i])!=fv)
			{
				values[i].execute(data);

				if(k!=keys[i])
				{
					++i;
				}
			}
		}

		if(data!=null && data is IEvt)
		{
			((IEvt)data).clear();
		}
	}

	public class BaseEventAction
	{
		public Action func;

		public Action<T> func2;

		public void execute(T data)
		{
			if(func!=null)
			{
				func();
				return;
			}

			if(func2!=null)
			{
				func2(data);
				return;
			}
		}
	}
}